Support computation of all well-quantized and vertical G4-fluxes

HereAround opened the pull request:
~~Sadly still WIP.~~

cc @apturner @emikelsons

HereAround commented:
Ok - this master piece (hopefully?) is finally getting ready. "What is in this PR?" is hear you ask:

* It implements the methods necessary to compute all well-quantized G4-fluxes. In particular, it adds specialized intersection theory, so we can handle the big model in a timely-fashion.
* I have serialized most of the computed data, so that we can save those fluxes in the .mrdi file of the big model. As such, once this big-model file is loaded, the data will be available almost instantly. Downside is that said .mrdi file is now almost 1GB in size.

I am waiting for the tests to pass now/will make adjustments so that they do.

Beyond this, there are the following improvements that are possible:
1. We could of course profile more to see if we get more performance. I have done some effort in this regard already, so this may not be the most pressing concern.
2. I wish to add a new type `FamilyOfG4Fluxes` which we can use to wrap the return value of  the new method `well_quantized_g4_fluxes`. Currently, it returns a tuple of two matrices and I added documentation to explain what this means. But then, this is not super cool for the end user. So it would be good to improve on this.

@apturner @emikelsons maybe we can discuss more shortly. If you find the time, take a (brief) look and let me know your thoughts.

HereAround commented:
This is ready for your reviews @apturner and @emikelsons .

HereAround commented:
@apturner @emikelsons please take a look at this PR.

emikelsons commented on experimental/FTheoryTools/src/G4Fluxes/special_attributes.jl:
```diff
@@ -262,3 +101,467 @@ function ambient_space_models_of_g4_fluxes(m::AbstractFTheoryModel; check::Bool
+
+      # Compute against zero section and base divisor
+      zsc = zero_section_class(m)
+      pos_zero_section = findfirst(x -> x == string(polynomial(zsc).f), string.(gS))
```
This can fail to find the index of the zero_section, even if it has been correctly specified

emikelsons replied:
`pos_zero_section = findfirst(x -> x == string(polynomial(zsc)), string.([polynomial(x) for x in cds]))`

emikelsons replied:
The above change should work

HereAround replied:
I am thinking about "vanilla scenarios": 5d toric ambient space `X_Sigma`, resolved 4d hypersurface `Y` therein and zero section given as the restriction of `V(c) subseteq X_Sigma` to `Y`, where `c` is a suitable homogeneous coordinate of the toric ambient space.

When you say "can fail", are you thinking of more general setups? E.g. setups in which the zero section is cut-out by a more complicated polynomial? Or what other types of failure are you thinking about?

And most important: Suggestion for improving @emikelsons ?

emikelsons replied:
Well, I've been testing this for the F11 toric hypersurface(hypersurface form), after specifying the zero_section_class "v" in the JSON file

emikelsons replied:
However, if the index of the zero section can be found, I have already tested the well_quantized_ambient_space_models_of_g4_fluxes for this model by taking a Z-linear combination of the output, where all coefficients are 1, and tested that with the other built-in method "passes_elementary_quantization_checks(g4)", which returns true

HereAround replied:
> I am thinking about "vanilla scenarios": 5d toric ambient space `X_Sigma`, resolved 4d hypersurface `Y` therein and zero section given as the restriction of `V(c) subseteq X_Sigma` to `Y`, where `c` is a suitable homogeneous coordinate of the toric ambient space.
> 
> When you say "can fail", are you thinking of more general setups? E.g. setups in which the zero section is cut-out by a more complicated polynomial? Or what other types of failure are you thinking about?
> 
> And most important: Suggestion for improving @emikelsons ?

Sorry - somehow part of your earlier messages was not displayed (to me) when writing this message. This was written as response after only seeing your first message "This can fail to find the index of the zero_section, even if it has been correctly specified", and is thus only partly relevant.

HereAround replied:
> `pos_zero_section = findfirst(x -> x == string(polynomial(zsc)), string.([polynomial(x) for x in cds]))`

Ok. I have a feeling what you are hinting at: `polynomial(zsc)` is an element of a quotient ring. The string of this object may depend on its representation, and you are suggesting to compare with the strings of the (polynomials of) the cohomology classes in `cds`, which are also elements of the same quotient ring. Hence, they both should be presented in the same way.

In contrast, I lifted the quotient ring element (with the `.f`) to a multivariate polynomial ring and compared with the generators of the Cox ring. I believe what you are hinting at is that after `.f`, the comparison may fail, because a different representation is chosen? Is that what is happening for you?

Either way - it feels like your suggestion may be the better way to go. Of course, we could also execute a comparison test in the quotient ring. But then, we know that this will not complete for the big model. So not point in doing this.

An alternative would be to introduce an attribute `pos_of_zero_section` or alike, that we specify for all models upon construction. But I did not want to go that heavy route. At least not just yet.

emikelsons replied:
You're right in the future we should have this attribute as we already have this information in the JSON file, but we lose it during construction

HereAround replied:
Maybe we can merge this one first, and then tend to this? Or we fix it in this PR.

Either way - I just pushed a code change with your suggested line to find the position of the zero section in case of a "generic" model (a.k.a. different from the big model).

HereAround commented:
@emikelsons and I found a bug in this PR yesterday, which we fixed yesterday. I have squashed all of those changes into one (hopefully) working commit. To verify that indeed things are working as intended, the tests have been extended accordingly. Also, I have extended on the documentation.

HereAround committed:
[FTheoryTools] Compute well-quantized and vertical G4-fluxes

HereAround commented:
I am confident that the tests pass once https://github.com/oscar-system/Oscar.jl/issues/4341 is resolved. @apturner and @emikelsons , could you please review this PR and tell me if it is good to go as is?

Certainly, this PR/change is not perfect. Among others, we could look for better performance. Add an attribute like `base_divisors`, `base_divisor_positions`, `zero_section_index` etc. Or we could have a check to tell if `c2` is even for a given FTheory models. Or we could also compute all G4-fluxes that do not break non-Abelian gauge groups. Or the names of these functions might change (I am very happy for suggestions).

Still - I have postponed these improvements to a point in the future.

HereAround commented:
Maybe we could add these points to our to-do list in https://github.com/oscar-system/Oscar.jl/issues/2699. And of course, we should gauge at some point, which of these tasks have been completed/need attention (and if so how urgent) etc...

emikelsons reviewed (approved):
Looks good and this has been extensively tested.
